src/delivery.cpp

```diff
@@ -0,0 +1,180 @@
+}
+
+void setProgressStatus(std::string status)
+{
+    std_msgs::String progressStatus;
+    std::stringstream ssProgressStatus;
+    ssProgressStatus << status;
+    progressStatus.data = ssProgressStatus.str();
+
+    progressStatusPub.publish(progressStatus);
+}
+
+bool findSetpoint(std::string location)
+{
+    if (location.compare("Location A") == 0)
+    {
+        assignSetpoint(placeA);
+    }
+    else if (location.compare("Location B") == 0)
+    {
+        assignSetpoint(placeB);
+    }
+    else if (location.compare("Location C") == 0)
+    {
+        assignSetpoint(placeC);
+    }
+    else if (location.compare("Location D") == 0)
+    {
+        assignSetpoint(placeD);
+    }
+    else
+    {
+        assignSetpoint(placeNil);
+        return 0;
+    }
+
+    return 1;
+}
+
+void senderLocationCallback(const std_msgs::String::ConstPtr &name)
+{
+    senderLocation = name->data.c_str();
+}
+
+void receiverLocationCallback(const std_msgs::String::ConstPtr &name)
+{
+    receiverLocation = name->data.c_str();
+}
+
+void reachedSetpointCallback(const std_msgs::Bool::ConstPtr &value)
+{   
+    reachedSetpointBool = value->data;
+}
+
+int main(int argc, char **argv)
+{
+    ros::init(argc,argv,"delivery");
+    ros::NodeHandle n;
+
+    // Publishers
+    availabilityPub = n.advertise<std_msgs::String>("availability",1000);
+    progressStatusPub = n.advertise<std_msgs::String>("progress",1000);
+    setpointPub = n.advertise<geometry_msgs::Vector3>("setpoint",1000);
+
+    // Subscribers
+    ros::Subscriber senderLocationSub = n.subscribe("senderLocation",1000,senderLocationCallback);
+    ros::Subscriber receiverLocationSub = n.subscribe("receiverLocation",1000,receiverLocationCallback);
+    ros::Subscriber reachedSetpointSub = n.subscribe("reachedSetpointBool",10,reachedSetpointCallback);
+
+    ros::Rate loopRate(10); 
+
+    while (ros::ok())
+    {
+        // Sender loop
+        int senderloopCount{0};
+        while (ros::ok())
+        {
+            bool foundSetpointBool{findSetpoint(senderLocation)};            
+            setpointPub.publish(setpointArray);
+
+            if (foundSetpointBool)
+            {
+                setAvailabilityStatus("no");
+                setProgressStatus("in progress");
+            }
+
+            if (senderloopCount++ == 0) {sleep(2);}
+            
+            ros::spinOnce();
+            loopRate.sleep();
+
+            if (reachedSetpointBool)
+            {
+                break;
+            }
+        }
+
+        // Receiver loop
+        int receiverloopCount{0};
+        while (ros::ok())
+        {
+            bool foundSetpointBool{findSetpoint(receiverLocation)};
+            setpointPub.publish(setpointArray);
+
+            if (receiverloopCount++ == 0) {sleep(2);}
+
+            ros::spinOnce();
+            loopRate.sleep();
+
+            if (reachedSetpointBool)
+            {
+                setProgressStatus("done");
```

Comment:
instead of a string, I would either create an enum of all possible options or use it in the method.
What if, a programmer, by mistake inputs 'done1' instead of done? How would you prevent this from happening?